File: delta2/layers/ops/kernels/framepow.cc
/* Copyright (C) 2017 Beijing Didi Infinity Technology and Development Co.,Ltd.
All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include "kernels/framepow.h"
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

namespace delta {
const float window_length_sec = 0.025;
const float frame_length_sec = 0.010;

FramePow::FramePow() {
  window_length_sec_ = window_length_sec;
  frame_length_sec_ = frame_length_sec;
  pf_FrmEng = NULL;
}

FramePow::~FramePow() { free(pf_FrmEng); }

void FramePow::set_window_length_sec(float window_length_sec) {
  window_length_sec_ = window_length_sec;
}

void FramePow::set_frame_length_sec(float frame_length_sec) {
  frame_length_sec_ = frame_length_sec;
}

int FramePow::init_eng(int input_size, float sample_rate) {
  f_SamRat = sample_rate;
  i_WinLen = static_cast<int>(window_length_sec_ * f_SamRat);
  i_FrmLen = static_cast<int>(frame_length_sec_ * f_SamRat);
  i_NumFrm = (input_size - i_WinLen) / i_FrmLen + 1;

  pf_FrmEng = static_cast<float*>(malloc(sizeof(float) * i_NumFrm));

  return 1;
}

int FramePow::proc_eng(const float* mic_buf) {
  int n, k;
  float* win = static_cast<float*>(malloc(sizeof(float) * i_WinLen));

  for (n = 0; n < i_NumFrm; n++) {
    pf_FrmEng[n] = 0.0;
    for (k = 0; k < i_WinLen; k++) {
      win[k] = mic_buf[n * i_FrmLen + k];
      pf_FrmEng[n] = pf_FrmEng[n] + win[k] * win[k];
    }
  }

  free(win);
  return 1;
}

int FramePow::get_eng(float* output) {
  memcpy(output, pf_FrmEng, sizeof(float) * i_NumFrm);

  return 1;
}

int FramePow::write_eng() {
  FILE* fp;
  fp = fopen("frame_energy.txt", "w");
  int n;
  for (n = 0; n < i_NumFrm; n++) {
    fprintf(fp, "%4.6f\n", pf_FrmEng[n]);
  }
  fclose(fp);
  return 1;
}
}  // namespace delta

File: delta/data/frontend/framepow.py
# Copyright (C) 2017 Beijing Didi Infinity Technology and Development Co.,Ltd.
# All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================
""""This model extracts framepow features per frame."""

import delta.compat as tf

from core.ops import py_x_ops
from delta.utils.hparam import HParams
from delta.data.frontend.base_frontend import BaseFrontend


class Framepow(BaseFrontend):
  """
  Compute power of every frame in speech. Return a float tensor with
  shape (1 * num_frames).
  """

  def __init__(self, config: dict):
    super().__init__(config)

  @classmethod
  def params(cls, config=None):
    """
    Set params.
    :param config: contains five optional parameters:
        --sample_rate       : Waveform data sample frequency (must match the waveform
                             file, if specified there). (float, default = 16000)
        --window_length		 : Window length in seconds. (float, default = 0.025)
        --frame_length		 : Hop length in seconds. (float, default = 0.010)
        --snip_edges			 : If True, the last frame (shorter than window_length)
                              will be cutoff. If False, 1 // 2 frame_length data will
                              be padded to data. (int, default = True)
        --remove_dc_offset : Subtract mean from waveform on each frame (bool, default = true)
    :return:An object of class HParams, which is a set of hyperparameters as name-value pairs.
    """

    window_length = 0.025
    frame_length = 0.010
    snip_edges = True
    remove_dc_offset = True
    sample_rate = 16000

    hparams = HParams(cls=cls)
    hparams.add_hparam('window_length', window_length)
    hparams.add_hparam('frame_length', frame_length)
    hparams.add_hparam('snip_edges', snip_edges)
    hparams.add_hparam('remove_dc_offset', remove_dc_offset)
    hparams.add_hparam('sample_rate', sample_rate)

    if config is not None:
      hparams.override_from_dict(config)

    return hparams

  def call(self, audio_data, sample_rate=None):
    """
    Caculate power of every frame in speech.
    :param audio_data: the audio signal from which to compute spectrum.
                       Should be an (1, N) tensor.
    :param sample_rate: [option]the samplerate of the signal we working with,
                        default is 16kHz.
    :return:A float tensor of size (1 * num_frames) containing power of every
            frame in speech.
    """

    p = self.config
    with tf.name_scope('framepow'):

      if sample_rate == None:
        sample_rate = tf.constant(p.sample_rate, dtype=tf.int32)

      assert_op = tf.assert_equal(
          tf.constant(p.sample_rate), tf.cast(sample_rate, dtype=tf.int32))
      with tf.control_dependencies([assert_op]):

        sample_rate = tf.cast(sample_rate, dtype=float)
        framepow = py_x_ops.frame_pow(
            audio_data,
            sample_rate,
            snip_edges=p.snip_edges,
            remove_dc_offset=p.remove_dc_offset,
            window_length=p.window_length,
            frame_length=p.frame_length)

        return tf.squeeze(framepow)
